### ISSUE TITLE
Processing the same signal twice yields different results

### ISSUE THREAD

**Issue by ilic-mezza** [facebookresearch/AudioDec#27]
Hi,

Thanks for the great work!

I was playing around with AudioDec and I consistently encountered the following problem.

When processing an audio clip for the first time, I get some results. When I process the same clip a second time, the discrete indices and the output signal differ from the ones that were produced the first time. After that, if I process the clip for a third time, results are consistent with the second pass.

From what I could see, the discrepancy seems to be limited to the very first few latent codes, and therefore only affects the initial part of the decoded signals. However, I expected a consistent behavior when encoding/decoding the same signal multiple times.

The mismatch occurs for all models I've tested, i.e., `vctk_v0`, `vctk_v1`, `vctk_v2`, `libritts_sym`, `libritts_v1`, regardless of the input signal. 

Here is a minimal example based on `demoFile.py` to reproduce the issue. 

```
import os
import torch
import librosa
import numpy as np
import matplotlib.pyplot as plt
from utils.audiodec import AudioDec, assign_model

# librosa example key
librosa_ex = "libri1"

# device assignment
tx_device = 'cpu'
rx_device = 'cpu'

# model assignment
sample_rate, encoder_checkpoint, decoder_checkpoint = assign_model("libritts_v1")

# AudioDec initinalize
print("AudioDec initinalizing!")
audiodec = AudioDec(tx_device=tx_device, rx_device=rx_device)
audiodec.load_transmitter(encoder_checkpoint)
audiodec.load_receiver(encoder_checkpoint, decoder_checkpoint)

with torch.no_grad():
    data, fs = librosa.load(librosa.ex(librosa_ex), sr=sample_rate, mono=True)

    data = data[:, None]  # (T,) -> (T,C), i.e., add channel dimension (last) 

    x = np.expand_dims(data.transpose(1, 0), axis=1)  # (T, C) -> (C, 1, T)

    x = torch.tensor(x, dtype=torch.float).to(tx_device)

    print("Encode/Decode...")
    z_1 = audiodec.tx_encoder.encode(x)
    idx_1 = audiodec.tx_encoder.quantize(z_1)
    zq_1 = audiodec.rx_encoder.lookup(idx_1)
    y_1 = audiodec.decoder.decode(zq_1)[:, :, :x.size(-1)]
    y_1 = y_1.squeeze(1).transpose(1, 0).cpu().numpy() # T x C

    print("Encode/Decode a second time...")
    z_2 = audiodec.tx_encoder.encode(x)
    idx_2 = audiodec.tx_encoder.quantize(z_2)
    zq_2 = audiodec.rx_encoder.lookup(idx_2)
    y_2 = audiodec.decoder.decode(zq_2)[:, :, :x.size(-1)]
    y_2 = y_2.squeeze(1).transpose(1, 0).cpu().numpy() # T x C

    print("Encode/Decode a third time...")
    z_3 = audiodec.tx_encoder.encode(x)
    idx_3 = audiodec.tx_encoder.quantize(z_3)
    zq_3 = audiodec.rx_encoder.lookup(idx_3)
    y_3 = audiodec.decoder.decode(zq_3)[:, :, :x.size(-1)]
    y_3 = y_3.squeeze(1).transpose(1, 0).cpu().numpy() # T x C

    # Test the decoded signals
    print(f'y_1 equals y_2? {np.all(y_1 == y_2)}')
    print(f'y_2 equals y_3? {np.all(y_2 == y_3)}')

    plt.figure()
    plt.plot(y_1 - y_2)
    plt.xlabel('Time index')
    plt.ylabel('y1 - y2')
    plt.title(f'{librosa_ex}')
    plt.tight_layout()
    plt.savefig(f'error_decoding_{librosa_ex}_y1-y2.png', dpi=300)
    plt.close()

    # Test the encoded indices
    print(f'idx_1 equals idx_2? {torch.all(idx_1 == idx_2)}')
    print(f'idx_2 equals idx_3? {torch.all(idx_2 == idx_3)}')

    num_errors = torch.count_nonzero(idx_1 != idx_2, dim=0).cpu().numpy()

    plt.figure()
    plt.stem(num_errors)
    plt.xlabel('Code time index')
    plt.ylabel('Num errors')
    plt.title(f'{librosa_ex}')
    plt.tight_layout()
    plt.savefig(f'error_decoding_{librosa_ex}_idx1-idx2.png', dpi=300)
    plt.close()

    plt.figure()
    plt.stem(num_errors[:40])
    plt.xlabel('Code time index')
    plt.ylabel('Num errors')
    plt.title(f'{librosa_ex} (close-up)')
    plt.tight_layout()
    plt.savefig(f'error_decoding_{librosa_ex}_idx1-idx2_zoom.png', dpi=300)
    plt.close()
```

Running the code above yields
```
$ y_1 equals y_2? False
$ y_2 equals y_3? True
```
```
$ idx_1 equals idx_2? False
$ idx_2 equals idx_3? True
```
Here are the plots obtained with `libritts_sym`
![error_decoding_libri1_y1-y2](https://github.com/facebookresearch/AudioDec/assets/56453983/c9d15408-f323-49d9-8b84-25c2f95fc7d2)
![error_decoding_libri1_idx1-idx2](https://github.com/facebookresearch/AudioDec/assets/56453983/1cd37b21-dba9-4c97-a171-36d8d246f1e6)
![error_decoding_libri1_idx1-idx2_zoom](https://github.com/facebookresearch/AudioDec/assets/56453983/4f26faca-14cc-4afc-9eed-55bb1d32af24)

The plots above are comparable for all models.

Similarly, I've noticed that the very first samples of the output signals differ when I encode an audio clip once,  do not modify the resulting discrete indices `idx` in any way, and decode `idx` twice.

Do you have any insight on this? 

Thanks!

**Comment by bigpon** [comment]
Hi,
did you use the causal version?
When using the causal version, each convolution/transpose convolution layer has a pad_buffer, which is initialized as zeros.
Since you encode/decode the same utterance three times, the pad_buffers are the same for the second and third encoding/decoding.
Therefore, you get the same results for the second and third coding.

The better way to code several utterances is to use the "reset_buffer(self)" before you code a new utterance.
However, since most utterances end in silence, there are almost no perceptual differences.

**Comment by ilic-mezza** [comment]
Hi @bigpon,

Thank you so much for your reply!

Yes, I am using the causal version. Somehow, I expected that resetting the pad_buffers after each call to `decoder.decode` would be the default behavior.

I agree, since most speech signals end with silence it shouldn't make much of a perceptual difference. However, I'm exploring the possibility of using AudioDec as a tokenizer. I'd be processing randomly-sampled short-time audio segments, which are likely not to end with silence. Therefore, the resulting codes would depend on the decoding of previous (randomly-chosen) segments.

I've tried to reset the buffers manually by adding the following lines every time I decode the utterance
```
audiodec.tx_encoder.reset_buffer()
audiodec.rx_encoder.reset_buffer()
audiodec.decoder.reset_buffer()
```
However, the problem persisted. (Am I doing something wrong here?)

The only way I managed to get consistent codes is by re-inizializing the modules via
```
audiodec.tx_encoder.initial_encoder(audiodec.receptive_length, tx_device)
_zq = audiodec.rx_encoder.initial_encoder(audiodec.receptive_length, rx_device)
audiodec.decoder.initial_decoder(_zq)
```
By doing so, I get the same codes (as shown below) but I still encounter a mismatch in the decoded signal (although negligible this time, see figure). 

```
$ y_1 equals y_2? False
$ y_2 equals y_3? True
$ idx_1 equals idx_2? True
$ idx_2 equals idx_3? True
```

Is there a more elegant way to make every call entirely independent of previous calls?


![error_decoding_libri1_y1-y2](https://github.com/facebookresearch/AudioDec/assets/56453983/414451cb-31d3-4c74-a3da-7dab481d29cc)

**Comment by bigpon** [comment]
Hi, 
the behaviors of "reset_buffer()" and "initial_encoder" are different.

"reset_buffer()" will make all padding buffers zeros.

"initial_encoder" will make the encoder encode a zero sequence (size= (1, self.input_channels, receptive_length)).
Since the convolution layers of the encoder have biases, the buffers are not zeros.

According to our previous experiments, using "initial_encoder" for real-time applications is safer to avoid any pop/artificial sound possibility.
